// @splitsoftware/splitio-react/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EXCEPTION_NO_REACT_OR_CREATECONTEXT = exports.ERROR_UTRACK_NO_USECONTEXT = exports.ERROR_UT_NO_USECONTEXT = exports.ERROR_UM_NO_USECONTEXT = exports.ERROR_UC_NO_USECONTEXT = exports.WARN_ST_NO_CLIENT = exports.ERROR_SC_NO_FACTORY = exports.ERROR_SF_NO_CONFIG_AND_FACTORY = exports.WARN_SF_CONFIG_AND_FACTORY = exports.getControlTreatmentsWithConfig = exports.CONTROL_WITH_CONFIG = exports.CONTROL = exports.OFF = exports.ON = exports.VERSION = void 0;
var utils_1 = require("./utils");
// The string below is a marker and will be replaced by the real version number. DO NOT CHANGE
exports.VERSION = 'react-' + '1.2.6';
// Treatments
exports.ON = 'on';
exports.OFF = 'off';
exports.CONTROL = 'control'; // SplitIO default value
exports.CONTROL_WITH_CONFIG = {
    treatment: 'control',
    config: null,
};
exports.getControlTreatmentsWithConfig = function (splitNames) {
    // validate split Names
    var validatedSplitNames = utils_1.validateSplits(splitNames);
    // return empty object if the returned value is false
    if (!validatedSplitNames)
        return {};
    // return control treatments for each validated split name
    return validatedSplitNames.reduce(function (pValue, cValue) {
        pValue[cValue] = exports.CONTROL_WITH_CONFIG;
        return pValue;
    }, {});
};
// Warning and error messages
exports.WARN_SF_CONFIG_AND_FACTORY = '[WARN] Both a config and factory props were provided to SplitFactory. Config prop will be ignored.';
exports.ERROR_SF_NO_CONFIG_AND_FACTORY = '[ERROR] SplitFactory must receive either a Split config or a Split factory as props.';
exports.ERROR_SC_NO_FACTORY = '[ERROR] SplitClient does not have access to a Split factory. This is because it is not inside the scope of a SplitFactory component or SplitFactory was not properly instantiated.';
exports.WARN_ST_NO_CLIENT = '[WARN] SplitTreatments does not have access to a Split client. This is because it is not inside the scope of a SplitFactory component or SplitFactory was not properly instantiated.';
var ERROR_NO_USECONTEXT = '[ERROR] Check your React version since `useContext` hook is not available. Split hooks require version 16.8.0+ of React.';
exports.ERROR_UC_NO_USECONTEXT = ERROR_NO_USECONTEXT + ' Returning null from `useClient` hook.';
exports.ERROR_UM_NO_USECONTEXT = ERROR_NO_USECONTEXT + ' Returning null from `useManager` hook.';
exports.ERROR_UT_NO_USECONTEXT = ERROR_NO_USECONTEXT + ' Returning control treatments from `useTreatments` hook.';
exports.ERROR_UTRACK_NO_USECONTEXT = ERROR_NO_USECONTEXT + ' Returning a no-op function from `useTrack` hook.';
exports.EXCEPTION_NO_REACT_OR_CREATECONTEXT = 'React library is not available or its version is not supported. Check that it is properly installed or imported. Split SDK requires version 16.3.0+ of React.';
