// @splitsoftware/splitio-react/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SplitContext = exports.useManager = exports.useTrack = exports.useTreatments = exports.useClient = exports.SplitFactory = exports.SplitClient = exports.SplitTreatments = exports.withSplitTreatments = exports.withSplitClient = exports.withSplitFactory = exports.SplitSdk = void 0;
// Split SDK factory (Renamed to avoid name conflict with SplitFactory component)
var splitio_1 = require("./splitio");
Object.defineProperty(exports, "SplitSdk", { enumerable: true, get: function () { return splitio_1.SplitFactory; } });
// HOC functions
var withSplitFactory_1 = require("./withSplitFactory");
Object.defineProperty(exports, "withSplitFactory", { enumerable: true, get: function () { return __importDefault(withSplitFactory_1).default; } });
var withSplitClient_1 = require("./withSplitClient");
Object.defineProperty(exports, "withSplitClient", { enumerable: true, get: function () { return __importDefault(withSplitClient_1).default; } });
var withSplitTreatments_1 = require("./withSplitTreatments");
Object.defineProperty(exports, "withSplitTreatments", { enumerable: true, get: function () { return __importDefault(withSplitTreatments_1).default; } });
// Render props components
var SplitTreatments_1 = require("./SplitTreatments");
Object.defineProperty(exports, "SplitTreatments", { enumerable: true, get: function () { return __importDefault(SplitTreatments_1).default; } });
var SplitClient_1 = require("./SplitClient");
Object.defineProperty(exports, "SplitClient", { enumerable: true, get: function () { return __importDefault(SplitClient_1).default; } });
var SplitFactory_1 = require("./SplitFactory");
Object.defineProperty(exports, "SplitFactory", { enumerable: true, get: function () { return __importDefault(SplitFactory_1).default; } });
// helper functions/hooks
var useClient_1 = require("./useClient");
Object.defineProperty(exports, "useClient", { enumerable: true, get: function () { return __importDefault(useClient_1).default; } });
var useTreatments_1 = require("./useTreatments");
Object.defineProperty(exports, "useTreatments", { enumerable: true, get: function () { return __importDefault(useTreatments_1).default; } });
var useTrack_1 = require("./useTrack");
Object.defineProperty(exports, "useTrack", { enumerable: true, get: function () { return __importDefault(useTrack_1).default; } });
var useManager_1 = require("./useManager");
Object.defineProperty(exports, "useManager", { enumerable: true, get: function () { return __importDefault(useManager_1).default; } });
// SplitContext
var SplitContext_1 = require("./SplitContext");
Object.defineProperty(exports, "SplitContext", { enumerable: true, get: function () { return __importDefault(SplitContext_1).default; } });
