// gatsby-plugin-use-dark-mode/theme-hydration-script-tag.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _terser = _interopRequireDefault(require("terser"));

// Adapted from:
// https://github.com/donavon/use-dark-mode/blob/develop/noflash.js.txt
function generateNoFlashScript({
  classNameDark,
  classNameLight,
  storageKey
}) {
  return `
    (function(classNameDark, classNameLight, storageKey) {
      function setClassOnDocumentBody(darkMode) {
        document.body.classList.add(darkMode ? classNameDark : classNameLight);
        document.body.classList.remove(darkMode ? classNameLight : classNameDark);
      }

      var preferDarkQuery = '(prefers-color-scheme: dark)';
      var mql = window.matchMedia(preferDarkQuery);
      var supportsColorSchemeQuery = mql.media === preferDarkQuery;
      var localStorageTheme = null;
      try {
        localStorageTheme = localStorage.getItem(storageKey);
      } catch (err) {}
      var localStorageExists = localStorageTheme !== null;
      if (localStorageExists) {
        localStorageTheme = JSON.parse(localStorageTheme);
      }

      // Determine the source of truth
      if (localStorageExists) {
        // source of truth from localStorage
        setClassOnDocumentBody(localStorageTheme);
      } else if (supportsColorSchemeQuery) {
        // source of truth from system
        setClassOnDocumentBody(mql.matches);
        localStorage.setItem(storageKey, mql.matches);
      } else {
        // source of truth from document.body
        var isDarkMode = document.body.classList.contains(classNameDark);
        localStorage.setItem(storageKey, JSON.stringify(isDarkMode));
      }
    })('${classNameDark}', '${classNameLight}', '${storageKey}');
  `;
}

function ThemeHydrationScriptTag({
  classNameDark,
  classNameLight,
  storageKey,
  minify
}) {
  const noFlashScript = generateNoFlashScript({
    classNameDark,
    classNameLight,
    storageKey
  });
  const finalNoFlashScript = minify ? _terser.default.minify(noFlashScript).code || '' : noFlashScript; // eslint-disable-next-line react/no-danger

  return /*#__PURE__*/_react.default.createElement("script", {
    dangerouslySetInnerHTML: {
      __html: finalNoFlashScript
    }
  });
}

ThemeHydrationScriptTag.propTypes = {
  classNameDark: _propTypes.default.string,
  classNameLight: _propTypes.default.string,
  storageKey: _propTypes.default.string,
  minify: _propTypes.default.bool
};
ThemeHydrationScriptTag.defaultProps = {
  classNameDark: 'dark-mode',
  classNameLight: 'light-mode',
  storageKey: 'darkMode',
  minify: true
};
var _default = ThemeHydrationScriptTag;
exports.default = _default;
