// @splitsoftware/splitio-react/es/splitio/storage/SegmentCache/InMemory/node.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import { numberIsInteger } from '../../../utils/lang';
import { _Set } from '../../../utils/lang/Sets';

var SegmentCacheInMemory = /*#__PURE__*/function () {
  function SegmentCacheInMemory(keys) {
    _classCallCheck(this, SegmentCacheInMemory);

    this.keys = keys;
    this.flush();
  }

  _createClass(SegmentCacheInMemory, [{
    key: "addToSegment",
    value: function addToSegment(segmentName, segmentKeys) {
      var values = this.segmentCache[segmentName];
      var keySet = values ? values : new _Set();
      segmentKeys.forEach(function (k) {
        return keySet.add(k);
      });
      this.segmentCache[segmentName] = keySet;
      return true;
    }
  }, {
    key: "removeFromSegment",
    value: function removeFromSegment(segmentName, segmentKeys) {
      var values = this.segmentCache[segmentName];
      var keySet = values ? values : new _Set();
      segmentKeys.forEach(function (k) {
        return keySet.delete(k);
      });
      this.segmentCache[segmentName] = keySet;
      return true;
    }
  }, {
    key: "isInSegment",
    value: function isInSegment(segmentName, key) {
      var segmentValues = this.segmentCache[segmentName];

      if (segmentValues) {
        return segmentValues.has(key);
      }

      return false;
    }
  }, {
    key: "registerSegment",
    value: function registerSegment(segmentName) {
      if (!this.segmentCache[segmentName]) {
        this.segmentCache[segmentName] = new _Set();
      }

      return true;
    }
  }, {
    key: "registerSegments",
    value: function registerSegments(segments) {
      for (var i = 0; i < segments.length; i++) {
        this.registerSegment(segments[i]);
      }

      return true;
    }
  }, {
    key: "getRegisteredSegments",
    value: function getRegisteredSegments() {
      return Object.keys(this.segmentCache);
    }
  }, {
    key: "setChangeNumber",
    value: function setChangeNumber(segmentName, changeNumber) {
      var segmentChangeNumberKey = this.keys.buildSegmentTillKey(segmentName);
      this.segmentChangeNumber[segmentChangeNumberKey] = changeNumber;
      return true;
    }
  }, {
    key: "getChangeNumber",
    value: function getChangeNumber(segmentName) {
      var segmentChangeNumberKey = this.keys.buildSegmentTillKey(segmentName);
      var value = this.segmentChangeNumber[segmentChangeNumberKey];
      return numberIsInteger(value) ? value : -1;
    }
  }, {
    key: "flush",
    value: function flush() {
      this.segmentCache = {};
      this.segmentChangeNumber = {};
    }
  }]);

  return SegmentCacheInMemory;
}();

export default SegmentCacheInMemory;
